Sources/NIOCore/SingleStepByteToMessageDecoder.swift

```diff
@@ -233,8 +233,9 @@ public final class NIOSingleStepByteToMessageProcessor<Decoder: NIOSingleStepByt
         }
 
         self._buffer = nil  // To avoid CoW
+        defer { self._buffer = buffer }
+
         let result = try body(&buffer)
```

Comment:
This is why we decided to move the `self._buffer = buffer` line into a `defer` block. This is actually what was triggering the original bug: `body` would throw but we'd never reinstate the original value of the buffer, so it would remain as `nil`, which caused issues when it happened reentrantly because the outer closure would come back to a context in which the buffer is now `nil`, and we were force-unwrapping it.
Rick mentioned this in https://github.com/apple/swift-nio/pull/2881#discussion_r1756862642

Reply:
🙈 I read the diff in reverse somehow haha. nice catch!